### ISSUE TITLE
Document that Ceph filestore will not work on zfs with "error -22"

### ISSUE THREAD

**Issue by IlyaSemenov** [rook/rook#779]
This comes from #427 where @travisn suggested that I open a new issue.

Running rook with `filestore` storage on a node where the file system for `dataDirHostPath` is zfs will fail with cryptic messages (error -22):

```console
$ kubectl logs rook-ceph-osd-c6s1l -n rook
2017-06-27 07:36:54.079774 I | exec: Running command: ceph osd create 46c1606b-1969-4f54-ab39-69ab4b7d219e --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/388692481
2017-06-27 07:36:54.339180 I | cephosd: successfully created OSD 46c1606b-1969-4f54-ab39-69ab4b7d219e with ID 4
2017-06-27 07:36:54.339464 I | cephmon: writing config file /var/lib/rook/osd4/rook.config
2017-06-27 07:36:54.339672 I | exec: Running command: ceph mon getmap --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/082305900
2017-06-27 07:36:54.477225 I | exec: got monmap epoch 4
2017-06-27 07:36:54.477423 I | cephosd: Initializing OSD 4 file system at /var/lib/rook/osd4...
2017-06-27 07:36:54.477625 I | proc: Running process ceph-osd with args: [--mkfs --mkkey --id=4 --cluster=rook --conf=/var/lib/rook/osd4/rook.config --osd-data=/var/lib/rook/osd4 --osd-uuid=46c1606b-1969-4f54-ab39-69ab4b7d219e --monmap=/var/lib/rook/osd4/tmp/activate.monmap --osd-journal=/var/lib/rook/osd4/journal --keyring=/var/lib/rook/osd4/keyring]
2017-06-27 07:36:54.477648 I | exec: Running command: ceph-osd --mkfs --mkkey --id=4 --cluster=rook --conf=/var/lib/rook/osd4/rook.config --osd-data=/var/lib/rook/osd4 --osd-uuid=46c1606b-1969-4f54-ab39-69ab4b7d219e --monmap=/var/lib/rook/osd4/tmp/activate.monmap --osd-journal=/var/lib/rook/osd4/journal --keyring=/var/lib/rook/osd4/keyring
2017-06-27 07:36:54.486629 I | mkfs-osd4: 2017-06-27 07:36:54.486495 7ff6e333cc80  0 ceph version 1677375ca (71677375ca98b267e07d99239194ba9f9c4b1889), process ceph-osd, pid 105
2017-06-27 07:36:54.501580 I | mkfs-osd4: 2017-06-27 07:36:54.501524 7ff6e333cc80  0 filestore(/var/lib/rook/osd4) backend generic (magic 0x2fc12fc1)
2017-06-27 07:36:54.520452 I | mkfs-osd4: 2017-06-27 07:36:54.520409 7ff6e333cc80 -1 filestore(/var/lib/rook/osd4) mkjournal error creating journal on /var/lib/rook/osd4/journal: (22) Invalid argument
2017-06-27 07:36:54.520552 I | mkfs-osd4: 2017-06-27 07:36:54.520442 7ff6e333cc80 -1 OSD::mkfs: ObjectStore::mkfs failed with error -22
2017-06-27 07:36:54.520560 I | mkfs-osd4: 2017-06-27 07:36:54.520523 7ff6e333cc80 -1  ** ERROR: error creating empty object store in /var/lib/rook/osd4: (22) Invalid argument
2017-06-27 07:36:54.521498 I | mkfs-osd4: 2017-06-27 07:36:54.520409 7ff6e333cc80 -1 filestore(/var/lib/rook/osd4) mkjournal error creating journal on /var/lib/rook/osd4/journal: (22) Invalid argument
2017-06-27 07:36:54.521503 I | mkfs-osd4: 2017-06-27 07:36:54.520442 7ff6e333cc80 -1 OSD::mkfs: ObjectStore::mkfs failed with error -22
2017-06-27 07:36:54.521506 I | mkfs-osd4: 2017-06-27 07:36:54.520523 7ff6e333cc80 -1  ** ERROR: error creating empty object store in /var/lib/rook/osd4: (22) Invalid argument
2017-06-27 07:36:54.521628 E | cephosd: failed to config osd in path /var/lib/rook. failed to initialize OSD at /var/lib/rook/osd4: failed osd mkfs for OSD ID 4, UUID 46c1606b-1969-4f54-ab39-69ab4b7d219e, dataDir /var/lib/rook/osd4: failed to run ceph-osd: Failed to complete mkfs-osd4: exit status 1
2017-06-27 07:36:54.521635 I | cephosd: 0/1 osd dirs succeeded on this node
failed to configure dirs map[/var/lib/rook:4]. failed to initialize OSD at /var/lib/rook/osd4: failed osd mkfs for OSD ID 4, UUID 46c1606b-1969-4f54-ab39-69ab4b7d219e, dataDir /var/lib/rook/osd4: failed to run ceph-osd: Failed to complete mkfs-osd4: exit status 1
```

I spinned the docker container manually and ran `strace`:

```console
# docker run --rm -it --entrypoint bash \
    -v /dev:/dev -v /srv/rook/data:/var/lib/rook
    quay.io/rook/rookd:master-latest
```

```console
# apt-get update && apt-get install -y strace
# trace ceph-osd --mkfs --mkkey --id=4 --cluster=rook \
    --conf=/var/lib/rook/osd4/rook.config --osd-data=/var/lib/rook/osd4 \
    --osd-uuid=46c1606b-1969-4f54-ab39-69ab4b7d219e \
    --monmap=/var/lib/rook/osd4/tmp/activate.monmap \
    --osd-journal=/var/lib/rook/osd4/journal --keyring=/var/lib/rook/osd4/keyring
...
open("/var/lib/rook/osd4/journal", O_RDWR|O_CREAT|O_DSYNC|O_DIRECT, 0644) = -1 EINVAL (Invalid argument)
futex(0x55e353f960bc, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x55e353f960b8, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x55e353f96038, FUTEX_WAKE_PRIVATE, 1) = 1
2017-06-27 07:53:21.912489 7f59464cdc80 -1 filestore(/var/lib/rook/osd4) mkjournal error creating journal on /var/lib/rook/osd4/journal: (22) Invalid argument
2017-06-27 07:53:21.912489 7f59464cdc80 -1 filestore(/var/lib/rook/osd4) mkjournal error creating journal on /var/lib/rook/osd4/journal: (22) Invalid argument
```

Searching for the `open()` flags revealed that zfs didn't support `O_DSYNC` (https://github.com/zfsonlinux/zfs/issues/224). In order to run rook on zfs, one needs to create a zvol and attach it with `useAllDevices: false` or `deviceFilter: zdXX`.

This is a non-trivial pitfal and it better be documented, for example in https://github.com/rook/rook/blob/master/Documentation/cluster-tpr.md in "Storage Configuration Settings" chapter.

**Comment by Queuecumber** [comment]
I was actually able to make this work without a zvol by including the following configmap in the manifest for my cluster:

```
kind: ConfigMap
apiVersion: v1
metadata:
  name: rook-config-override
  namespace: storage
data:
  config: |
    [osd]
    journal dio = false
```

This disables the journal direct write. I don't know if this is a bad idea though

**Comment by Amos-85** [comment]
Hi @IlyaSemenov,
you'll need to separate the osd to empty host directory that not sit on zfs dataset,
there is no issue the monitor data will be on zfs dataset,
if you don't set osd directory, the osd store object will mount on the same rook directory tree.

something similar to that configuration

```
apiVersion: ceph.rook.io/v1beta1
kind: Cluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  dataDirHostPath: </rook_mon_data_dir>     ####CEPH MON DIR
  serviceAccount: rook-ceph-cluster
  network:
    hostNetwork: false
  dashboard:
    enabled: true
  # cluster level storage configuration and selection
  storage:
    useAllNodes: true
    useAllDevices: false
    deviceFilter:
    location:
    directories:
    - path: </ceph-osd_dir>          ####CEPH OSD DIR   ### Can be also override per node
    nodes: 
    - name: "<SomeHost>"
      devices:
      - name: "sdf"
      - name: "sdg"
      - name: "sdh"
      - name: "sdi"
```
I'm also using zfs on my servers and this configuration works for me,
Cheers.

**Comment by IlyaSemenov** [comment]
On an another note, https://github.com/zfsonlinux/zfs/issues/224 has been closed a couple months ago (via https://github.com/zfsonlinux/zfs/pull/7823) and Direct IO is told to be available in zfs 0.8. Not sure if it will be directly supported by Cephs then.
